### PR TITLE
Implement Authentication Method Reference (AMR) claim from OIDC speci…

### LINKED ISSUE #19190: Add "amr" to already implemented "acr" support
### Description

Keycloak already has support for **Authentication Context Class Reference (ACR)**.

Missing support for **Authentication Methods References** to be implemented in accordance to RFC-8176.

### Discussion

_No response_

### Motivation

It currently is possible to check the **Level of Authentication** and request the correct level when required.
Unfortunately it is not possible to assure which combination of methods have been used, nor is it possible to request which methods are allowed/required.

This is part of the [OpenID Connect Core](https://openid.net/specs/openid-connect-core-1_0.html) (ID Token) specification.

### Details

AMR [RFC-8176](https://www.rfc-editor.org/rfc/rfc8176.html)
LoA [RFC-6711](https://www.rfc-editor.org/rfc/rfc6711.html)

[Proper support for 'acr' and 'amr'](https://issues.redhat.com/browse/KEYCLOAK-3314)

### PR DESCRIPTION
**Pull request by ben95cd**
Implement Authentication Method Reference (AMR) claim from OIDC specification

This implements a method for configuring authenticator reference values for Keycloak authenticator executions and a protocol mapper for populating the AMR claim in the resulting OIDC tokens.

This implementation adds a default configuration item to each authenticator execution, allowing administrators to configure an authenticator reference value. Upon successful completion of an authenticator during an authentication flow, Keycloak tracks the execution ID in a user session note.

The protocol mapper pulls the list of completed authenticators from the user session notes and loads the associated configurations for each authenticator execution. It then captures the list of authenticator references from these configs and sets it in the AMR claim of the resulting tokens.

Closes #19190

### CODE REVIEW

**Review by pedroigor — DISMISSED** [review]
Thanks, LGTM.

I'm only not sure if we want to enable the mapper by default to clients in a realm. Or if we want migration steps to add this mapper to existing realms (probably not).

@mposolda Wdyt?

File: js/apps/admin-ui/src/authentication/components/ExecutionConfigModal.tsx [hunk 1]
```diff
@@ -54,22 +54,52 @@ export const ExecutionConfigModal = ({
...
     async () => {
-      const configDescription =
-        await adminClient.authenticationManagement.getConfigDescription({
-          providerId: execution.providerId!,
-        });
       let config: AuthenticatorConfigRepresentation | undefined;
+      let configDescription: AuthenticatorConfigInfoRepresentation;
+
+      // fetch execution config if its configurable
+      if (execution.configurable) {
```
Comment (jonkoops) [inline js/apps/admin-ui/src/authentication/components/ExecutionConfigModal.tsx:79]: If you need to assign a value with two branches, you can use a ternary operator to differentiate them and do a direct assignment.

```ts
const foo = condition ? a : b;
```

File: js/apps/admin-ui/src/authentication/components/FlowRow.tsx [hunk 2]
```diff
@@ -102,9 +102,7 @@ export const FlowRow = ({
                 />
               </DataListCell>,
               <DataListCell key={`${execution.id}-config`}>
-                {execution.configurable && (
```
Comment (jonkoops) [inline js/apps/admin-ui/src/authentication/components/FlowRow.tsx:105]: Are you sure this condition can be removed? Or is this some debugging code?

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Review by mposolda — COMMENTED** [review]
@ben95cd Thanks for the contribution. Few questions:

- In current Keycloak, only authentication flows with `AuthenticatorFactory.isConfigurable()` have the shown link to the "configuration" in the admin console. If I understand correctly, with your changes every authenticator has shown the "configuration" and have possibility to have "Authentication reference" configured. Is it correct? (Sorry for asking as I did not try your changes myself yet)

- Is there ability to define also `amr` on the subflow level? I can see for instance in the specification https://www.rfc-editor.org/rfc/rfc8176.html#page-5 that there is `mfa` claim to specify multifactor authentication. So possibly people may want to just specify that multifactor authentication was used (without requirement to specify if `otp` or `webauthn` or different method was used as concrete multifactor mechanism). Hence it can be useful to add the `mfa` directly to the multifactor authentication subflow. Not sure if it is blocker for this PR or perhaps something to do as a follow-up? WDYT @pedroigor ?

- I am not sure about SSO authentication cases together with this. For instance the example:
-- User authenticates with password+otp. The `acr` claim will contain level 2 and `amr` will contain both `password` and `otp`.
-- Now user re-authenticates by SSO. Now should the `amr` claim be present with `password` + `otp` or not? Since this request was just SSO re-authentication and not "active" authentication with password+otp, it is not clear to me whether it should be always present.

IMO in case that step-up authentication is used, it can be good to align with the same like we do for `acr` claim.  It looks strange to me return `acr=0` and at the same time, return `amr` with all the values from userSession notes, which were ever used for this user session.

For instance level `acr=2` is used just if ACR authentication happened in current request or in relatively short time before (there is `max age` possibility to be specified on the "Loa conditions" for this purpose). Could we somehow re-use this ability? In relation to this, I am not 100% sure if using simple userSession notes is enough for amr... For example step-up authentication currently uses some more context info like authTime etc. See for example `AcrStore.getCurrentAuthenticatedLevelsMap`

**Comment by mposolda** [comment]
> Thanks, LGTM.
> 
> I'm only not sure if we want to enable the mapper by default to clients in a realm. Or if we want migration steps to add this mapper to existing realms (probably not).
> 
> @mposolda Wdyt?

My vote is to not add anything to existing realms and not do anything for the migration. But possibly have this protocol mapper defined by default in new realms. Maybe in the same client scope where the `acr` protocol mapper is defined as this is related to that.

File: js/apps/admin-ui/src/authentication/components/FlowRow.tsx (hunk 2 above)
Comment (ben95cd) [reply js/apps/admin-ui/src/authentication/components/FlowRow.tsx:105]: Hi @jonkoops, this was originally used to hide the execution configuration if the authenticator is not marked as configurable. however, this implementation sets a default configuration on all authenticators for the authenticator reference, so this check was moved to inside the ExecutionConfigModal component to line 77, and will fetch the config description from keycloak if the execution is set as configurable or default to an empty list, then add the default authenticator reference config to that list of configuration items.

**Comment by ben95cd** [comment]
> In current Keycloak, only authentication flows with AuthenticatorFactory.isConfigurable() have the shown link to the "configuration" in the admin console. If I understand correctly, with your changes every authenticator has shown the "configuration" and have possibility to have "Authentication reference" configured. Is it correct? (Sorry for asking as I did not try your changes myself yet)

Yes, you're correct thats how it works.

> Is there ability to define also amr on the subflow level? I can see for instance in the specification https://www.rfc-editor.org/rfc/rfc8176.html#page-5 that there is mfa claim to specify multifactor authentication. So possibly people may want to just specify that multifactor authentication was used (without requirement to specify if otp or webauthn or different method was used as concrete multifactor mechanism). Hence it can be useful to add the mfa directly to the multifactor authentication subflow. Not sure if it is blocker for this PR or perhaps something to do as a follow-up? WDYT @pedroigor ?

That would be an easy change to make, you could still achieve that goal with the existing implementation by assigning all authenticators in that sub-flow the same reference value. Your suggestion would definitely be a cleaner way to accomplish that goal though. Happy to add it to the PR if you'd like.

> IMO in case that step-up authentication is used, it can be good to align with the same like we do for acr claim. It looks strange to me return acr=0 and at the same time, return amr with all the values from userSession notes, which were ever used for this user session.

That is a great point, I would definitely agree with you. I can update the PR and implement a similar time bound way to determine if the amr values are still valid.

File: js/apps/admin-ui/src/authentication/components/ExecutionConfigModal.tsx (hunk 1 above)
Comment (ben95cd) [reply js/apps/admin-ui/src/authentication/components/ExecutionConfigModal.tsx:79]: Thats much cleaner, i will update that when i update the PR based on @mposolda 's feedback below

File: services/src/main/java/org/keycloak/protocol/oidc/mappers/AmrProtocolMapper.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+ * completed authenticators found in the user session notes.
+ */
+public class AmrProtocolMapper extends AbstractOIDCProtocolMapper implements OIDCAccessTokenMapper, OIDCIDTokenMapper, EnvironmentDependentProviderFactory {
+
+    private static final Logger logger = Logger.getLogger(AmrProtocolMapper.class);
+
+    private static final List<ProviderConfigProperty> configProperties = new ArrayList<>();
+
+    static {
+        OIDCAttributeMapperHelper.addIncludeInTokensConfig(configProperties, AmrProtocolMapper.class);
```
[the remark this answers is not in the document]
Comment (ben95cd) [reply services/src/main/java/org/keycloak/protocol/oidc/mappers/AmrProtocolMapper.java:53]: Will do. I'll post the updates when I update the PR based on @mposolda 's feedback below.

**Comment by ben95cd** [comment]
@pedroigor @mposolda @jonkoops I updated the PR with the changes and feedback that you all gave. Just let me know if there is anything else you'd like me to change.

**Review by pedroigor — COMMENTED** [review]
@mposolda I agree with all your points.

**Comment by pedroigor** [comment]
@ben95cd Re-running failing tests as they don't seem related.

**Comment by ben95cd** [comment]
Hi @pedroigor, i'm just checking in to see if there were any additional comments you had on my changes to address the comments above.

**Comment by jonkoops** [comment]
@ben95cd side-note, could you rebase the PR on `main` and fix the merge conflicts? I'd be more that happy to give this another review pass.

**Comment by ben95cd** [comment]
Thanks @jonkoops! I just fixed the merge conflict and rebased the PR.

**Comment by pedroigor** [comment]
@ben95cd Now we need a sign-off commit message and for that we probably want to squash all commits to make it easier :)

For that, create a commit by passing the `-s` such as `git commit -s ...`.

**Comment by ben95cd** [comment]
@pedroigor sorry about that! I added the signed off message and squashed the commits.

**Comment by jonkoops** [comment]
@andymunro could I ask you to vet the changes made to the documentation here?

**Review by jonkoops — APPROVED** [review]
Front-end changes look okay to me.

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@ben95cd Thanks!

@mposolda There are a few nitpicks (e.g.: start imports) which we can handle as a follow-up. The mapper is also not enabled by default to new realms and we can evaluate this behavior in a follow-up.

**Comment by ben95cd** [comment]
Thanks for merging this in!

**Pull request merged**